nova/objects/instance.py

```diff
@@ -1697,6 +1709,22 @@ def _get_counts_in_db_baremetalaware(context, project_id, user_id=None):
 
         return counts
 
+    @staticmethod
+    @db.pick_context_manager_reader
+    def _get_counts_in_db_by_project_and_az(context, project_id=None):
+        query = InstanceList._build_instance_usage_query(
+            context, project_id,
+            models.Instance.project_id,
+            models.Instance.availability_zone)
+
+        result = query.group_by(models.Instance.project_id,
+                                models.Instance.availability_zone).all()
```

Comment:
Does this overwrite the exiting `group_by()` or add to it? From the docs of SQLAlchemy this is not clear to me, but I would assume it replaces the existing ones. This would mean, we need to add the `models.Instance.instance_type_id` here, too.

Would it maybe make sense to supporting passing a `additional_group_by` kwarg in `_build_instance_usage_query()`?

Reply:
> Does this overwrite the exiting `group_by()` or add to it?

It adds, otherwise I wouldn't get the grouping by `instance_type_id`. 
The query object is build by method chaining, and there is no difference between:
```
def build_query(...):
     return query.function()
a = build_query().
```
vs
```
def build_query(...):
     return query

a = build_query().function()
```

I didn't know how to do the same with the fields (or filters).

But I found [Query.add_columns](https://docs.sqlalchemy.org/en/14/orm/query.html#sqlalchemy.orm.Query.add_columns), so I'll check if I can move that out.

Reply:
I moved now all the extra stuff out of the `_build_instance_usage_query()` method, and add modify the query for my needs.